Improve multi-controller support in HID and Controller Applet

mageven opened the pull request:
Feature changes:
- HID now allows reconfiguring guest controllers during runtime! No need to restart emulation.
- Controller Applet now supports multiple controllers
- Controller Applet now displays a Message dialog indicating config issues if any, and waits for user to correct them.
- HID now ignores Handheld config when Docked mode is enabled
- Auto-reassignment is dialed back (Wasn't assigning valid configs most of the time anyway) to rely more on Controller Applet.
- Fun change: Adds official LED patterns to Player IDs
- Misc change: Hotkey toggles now only work when Emu window is focused
- **NEW** Controller Setup Window now only shows valid options for type (For e.g., prevents setting up Pro Controller to Handheld)

Code changes:
- Rewrote NpadDevices to be consistent with other devices. Simplified approach with Update loop managing configuration.
- HID also considers NpadIdType now
- Due to above, `SetSupportedNpadXX` calls are ultra low cost now
- **NEW** Replaced HID specific `StructArrayHelpers` with the one in `Common.Memory`

No pictures and videos this time. Too many changes. Sorry about that.

Resolves #1448 
Resolves #1149 
Resolves #1296 

**Disclaimer:** This doesn't fix all the problems that HID & Applet have. Just a few steps forward.

**Note:** While I'm sure there's going to be plenty of good suggestions and changes that can be made, I'm unable to touch more code (like wrestling with GTK) atm unless there's a bug/regression.

gdkchan commented on Ryujinx.HLE/HOS/Services/Hid/HidDevices/NpadDevices.cs:
```diff
@@ -81,77 +110,87 @@ public void AddControllers(params ControllerConfig[] configs)
+        }
 
-            MatchControllers();
+        public void Update(List<GamepadInput> states)
```
Sounds like this could just take a `IEnumerable<GamepadInput>`. There's nothing on the code that requires modifying this list or directly indexing the elements, so using a foreach instead of the for loop below should be fine, and doing the change make `states` immutable which is best pratice.

mageven commented:
Replies in single comment.

> I'll make the same comment I made for swkbd here. I don't think that the HLE project should be making assumptions of how the GUI wants the parameters formatted. It should just pass a raw struct with the parameters provided by the game, and the GUI can figure out what to do with it.

Yes, this was a last minute change and I forgot I was using GTK markup. Fixed.

> Sounds like this could just take a IEnumerable<GamepadInput>. There's nothing on the code that requires modifying this list or directly indexing the elements, so using a foreach instead of the for loop below should be fine, and doing the change make states immutable which is best pratice.

There's no good equivalent interface for a read-only indexable storage, so for now I've used `IList`.

Rebased. Also removed HID specific `StructArrayHelpers` since there's one in `Common.Memory`

AcK77 commented on Ryujinx.HLE/HOS/Services/Hid/HidDevices/NpadDevices.cs:
```diff
@@ -1,74 +1,102 @@
-            _styleSetUpdateEvents = new KEvent[_maxControllers];
+            for (int i = 0; i < MaxControllers; ++i)
+            {
+                var npad = _configuredTypes[i];
```
Could you use an explicit cast here ?

AcK77 commented on Ryujinx.HLE/HOS/Services/Hid/HidDevices/NpadDevices.cs:
```diff
@@ -1,74 +1,102 @@
+                    continue;
+                }
+
+                var currentType = _device.Hid.SharedMemory.Npads[i].Header.Type;
```
Could you use an explicit cast here ?

AcK77 commented on Ryujinx.HLE/HOS/Services/Hid/HidDevices/NpadDevices.cs:
```diff
@@ -81,77 +109,87 @@ public void AddControllers(params ControllerConfig[] configs)
+        private void Remap()
         {
-            PrimaryController = PlayerIndex.Unknown;
+            var configs = _configuredTypes;
```
Could you use an explicit cast here ?

AcK77 commented on Ryujinx.HLE/HOS/Services/Hid/HidDevices/NpadDevices.cs:
```diff
@@ -81,77 +109,87 @@ public void AddControllers(params ControllerConfig[] configs)
+            for (int i = 0; i < MaxControllers; ++i)
             {
-                ref NpadConfig config = ref _configuredNpads[i];
+                var config = configs[i];
```
Could you use an explicit cast here ?

AcK77 commented on Ryujinx.HLE/HOS/Services/Hid/HidDevices/NpadDevices.cs:
```diff
@@ -81,77 +109,87 @@ public void AddControllers(params ControllerConfig[] configs)
-        private void InitController(PlayerIndex player, ControllerType type)
-        {
-            if (type == ControllerType.Handheld)
+            var oldType = controller.Header.Type;
```
Could you use an explicit cast here ?

AcK77 commented on Ryujinx.HLE/HOS/Services/Hid/HidDevices/NpadDevices.cs:
```diff
@@ -81,77 +109,87 @@ public void AddControllers(params ControllerConfig[] configs)
+                Logger.Info?.Print(LogClass.Hid, $"Disconnected Controller {oldType} from {player}");
+                _styleSetUpdateEvents[(int)player].ReadableEvent.Signal(); // signal disconnect
+                _activeCount--;
+                return;
```
Add an empty line before

AcK77 commented on Ryujinx/Ui/ControllerWindow.cs:
```diff
@@ -894,24 +922,31 @@ private void SaveToggle_Activated(object sender, EventArgs args)
             }
 
+            // Atomically replace and signal input change
+            // Note: Do not modify InputConfig.Value directly as then events aren't signaled
```
```suggestion
            // NOTE: Do not modify InputConfig.Value directly as then events aren't signaled.
```

AcK77 commented on Ryujinx/Ui/ControllerWindow.cs:
```diff
@@ -894,24 +922,31 @@ private void SaveToggle_Activated(object sender, EventArgs args)
                 }
             }
 
+            // Atomically replace and signal input change
```
```suggestion
            // Atomically replace and signal input change.
```

AcK77 commented on Ryujinx/Ui/ControllerWindow.cs:
```diff
@@ -91,6 +92,23 @@ private ControllerWindow(Builder builder, PlayerIndex controllerId, VirtualFileS
+                _controllerType.Append(ControllerType.JoyconRight.ToString(), "Joycon Right");
+            }
+
+            _controllerType.Active = 0; // set initial value to first in list
```
```suggestion
            _controllerType.Active = 0; // Set initial value to first in list.
```

Thog commented:
Rebase required

mageven commented:
> Rebase required

Done

AcK77 reviewed (approved):
Lgtm! Thanks for fixing that! Pretty nice job!

Thog commented:
Seems to not be working on MK8D with two players in split screen: the game freezes at the player selection.
